### DISCUSSION TITLE
Source environment variables to make them available from within devcontainer.json

### DISCUSSION THREAD

**Discussion by asperry** [devcontainers/community#119] [Q&A]
I understand that in a `devcontainer.json` file, I can use `${localEnv:VARIABLE_NAME}` to access environment variables on the host. This is very useful, but I would like to have access to some project-specific variables that I normally access by sourcing a bash script from the project. Is there a way to effectively source a bash script to modify the environment variables before the `devcontainer.json` gets parsed such that they are available through the `${localEnv:VARIABLE_NAME}` substitution mechanism?

As an example, one of these project-specific variables defines the docker image/tag string for the project. I'd like to use this value for the `image` property in the `devcontainer.json` file so that I can have a single source of truth for the docker image/tag string in my project.

I'm aware that I can achieve what I'm after by sourcing my project-specific script before running my `devcontainer.json` supporting tool (in this case, I'm using VS Code). But I'd prefer a solution that does not place constraints on how the user starts the tool.

**Comment by samruddhikhandale** [comment]
Hi 👋 

Looping in @chrmarti to help answer your question. 

On a side note, if you use supporting tools like GitHub Codespaces then you'd be able to configure repo secrets to achieve this. See https://docs.github.com/en/codespaces/managing-your-codespaces/managing-your-account-specific-secrets-for-github-codespaces

**Reply by chrmarti** [reply]
Not sure if a static env file would cover your case, but there is a feature request for supporting `.env` files: https://github.com/microsoft/vscode-remote-release/issues/3007.

**Reply by asperry** [reply]
Thank you for your response, but this looks like it is a technique for setting environment variables inside the container which is not my goal. I want to be able to set some local environment variables on the host such that they are available when parsing the `devcontainer.json` file before the development container is created or started. This may be a feature who's responsibility rests on the `devcontainer.json` supporting tool rather than on `devcontainers` itself perhaps?

**Comment by nataziel** [comment]
I would love if you could dynamically set variables/build args like so:
```
    "build": {
        // Sets the run context to one level up instead of the .devcontainer folder.
        "context": "..",
        "dockerfile": "Dockerfile",
        "args": {
            "PIP_EXTRA_INDEX_URL": "$(cat ~/.config/config.txt)"
        }
    },
```

**Comment by alec-flowers** [comment]
I would find this very useful. Right now it feels the localEnv behavior is still suspect at times in picking up variables. It shouldn't be the only way that we can dynamically specify mount points. 

For example if I want to set a mount point based on a localEnv, but some users may or may not set it, it would be nice to be able to set a fallback that is also dynamic. For example:

"source=${localEnv:HF_HOME:${localEnv:HOME}/.cache/huggingface},target=/home/ubuntu/.cache/huggingface,type=bind"

However such a construct is not supported. 
@chrmarti
